### PR TITLE
fix 5-arg `mul!` for vectors of vectors

### PR DESCRIPTION
**Pull request by ranocha**
On current `master` and release, vectors of vectors work with `*` and 3-arg `mul!` but not with 5-arg `mul!`:

```julia
julia> using LinearAlgebra

julia> u = [[1, 2], [3, 4]]
2-element Vector{Vector{Int64}}:
 [1, 2]
 [3, 4]

julia> A = [1 2; 3 4]
2×2 Matrix{Int64}:
 1  2
 3  4

julia> A * u
2-element Vector{Vector{Int64}}:
 [7, 10]
 [15, 22]

julia> v = [[0, 0], [0, 0]]
2-element Vector{Vector{Int64}}:
 [0, 0]
 [0, 0]

julia> mul!(v, A, u)
2-element Vector{Vector{Int64}}:
 [7, 10]
 [15, 22]

julia> mul!(v, A, u, 1, 1)
ERROR: MethodError: no method matching +(::Vector{Int64}, ::Int64)
For element-wise addition, use broadcasting with dot syntax: array .+ scalar

Closest candidates are:
  +(::Any, ::Any, ::Any, ::Any...)
   @ Base operators.jl:578
  +(::T, ::T) where T<:Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8}
   @ Base int.jl:87
  +(::Base.TwicePrecision, ::Number)
   @ Base twiceprecision.jl:290
  ...

Stacktrace:
 [1] MulAddMul
   @ ~/.../julia/stdlib/v1.10/LinearAlgebra/src/generic.jl:58 [inlined]
 [2] generic_matvecmul!(C::Vector{Vector{Int64}}, tA::Char, A::Matrix{Int64}, B::Vector{Vector{Int64}}, _add::LinearAlgebra.MulAddMul{true, false, Int64, Int64})
   @ LinearAlgebra ~/.../julia/stdlib/v1.10/LinearAlgebra/src/matmul.jl:810
 [3] mul!(y::Vector{Vector{Int64}}, A::Matrix{Int64}, x::Vector{Vector{Int64}}, alpha::Int64, beta::Int64)
   @ LinearAlgebra ~/.../julia/stdlib/v1.10/LinearAlgebra/src/matmul.jl:81
 [4] top-level scope
   @ REPL[7]:1
```

Since matrix vector products only need to be able to
- multiply elements of the matrix and the vector
- add elements of the vector multiplied by entries of the matrix

this is an unnecessary restriction I fixed here. In practice, this appears for us when multiplying vectors of `SVector`s from StaticArrays.jl (or `StructArray`s thereof from StructArrays.jl) by matrices, e.g., when discretizing differential equations.

### CODE REVIEW

File: stdlib/LinearAlgebra/src/matmul.jl [hunk 1]
```diff
@@ -807,7 +807,15 @@ function generic_matvecmul!(C::AbstractVector{R}, tA, A::AbstractVecOrMat, B::Ab
...
-            b = _add(B[k], false)
+            bk = B[k]
+            # We use `false` for `Number`s since `Bool`s are also numbers and
+            # we assume that we can add them. This should be more efficient in
+            # general since it avoids allocating a zero number for something
+            # like `BigFloats`. However, we do not necessarily need to have
+            # vectors of numbers but can also have vectors of vectors. In this
+            # case, adding a scalar `false` to the vector-valued entries is
+            # not defined, so we need to use an appropriate zero.
+            b = _add(bk, bk isa Number ? false : zero(bk))
```
Comment (N5N3) [inline stdlib/LinearAlgebra/src/matmul.jl:818]:
```suggestion
            b = _add(B[k])
```
Should be enough.

**Comment by dkarrasch** [comment]
What happens if `u` is a matrix of vectors? Does it also require a fix?

**Comment by ranocha** [comment]
> What happens if `u` is a matrix of vectors? Does it also require a fix?

I added additional tests with matrices of matrices - they work correctly in this PR.

**Comment by ranocha** [comment]
The only failing test is "test x86_64-w64-mingw32" with message "Received cancellation signal, interrupting", which does not seem to be related to this PR.

**Comment by brenhinkeller** [comment]
Looks like that failing test is just #47710 which should be fixed by #47711, so seems good to go here

**Pull request merged**